### PR TITLE
add blockfetcher to consensus

### PR DESCRIPTION
**Pull request by shivdeep-singh-ibm**
Signed-off-by: Shivdeep Singh <Shivdeep.Singh@ibm.com>

#### Type of change

- New feature

### CODE REVIEW

File: orderer/consensus/etcdraft/blockpuller.go [hunk 1]
```diff
@@ -107,3 +109,115 @@ func NewBlockPuller(support consensus.ConsenterSupport,
...
+	return &LedgerBlockPuller{
+		Height:         support.Height,
+		BlockRetriever: support,
+		BlockPuller:    &bf,
+	}, nil
+}
+
+func isBehaviourSuspicious(attestations []*orderer.BlockAttestation) bool {
+	// function to compare attestations
+	equal := func(x, y *orderer.BlockAttestation) bool {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/blockpuller.go:198]: only header and number? what about the data hash?

File: orderer/consensus/etcdraft/blockpuller.go [hunk 2]
```diff
@@ -107,3 +109,115 @@ func NewBlockPuller(support consensus.ConsenterSupport,
...
+		if x.Header.Number != y.Header.Number || len(x.Metadata.Metadata) != len(y.Metadata.Metadata) {
+			return false
+		}
+		// TODO: metadata can be unmarshalled
+		// and compared
+		return true
+	}
+
+	if attestations == nil || len(attestations) == 1 && attestations != nil {
```
[the remark this answers is not in the document]
Comment (shivdeep-singh-ibm) [reply orderer/consensus/etcdraft/blockpuller.go:207]: it is to handle case when 
when attestation pointer is nil
_a) attestation = nil_  

or there is only one attestation, which is nil
_b) attestations = [ nil ]_

[the remark this answers is not in the document]
Comment (yacovm) [reply orderer/consensus/etcdraft/blockpuller.go:207]: i get that, but - what are we trying to accomplish here? why are you doing this? what is the semantic meaning behind these conditions?

File: orderer/consensus/etcdraft/blockpuller.go [hunk 3]
```diff
@@ -107,3 +119,108 @@ func NewBlockPuller(support consensus.ConsenterSupport,
...
+			}
+		},
+		Config:            fc,
+		Logger:            flogging.MustGetLogger("orderer.common.cluster.puller").With("channel", support.ChannelID()),
+		ShuffleTimeout:    shuffleTimeout,
+		LastShuffledAt:    time.Now(),
+		MaxByzantineNodes: maxByzantineNodes,
+		ConfirmByzantineBehavior: func(attestations []*orderer.BlockAttestation) bool {
+			if attestations == nil {
+				return true
```
Comment (yacovm) [inline orderer/consensus/etcdraft/blockpuller.go:198]: why true? I thought we return whether we detected byzantine behavior... if we have no attestations then why do we return true?

File: orderer/consensus/etcdraft/blockpuller.go [hunk 4]
```diff
@@ -107,3 +119,108 @@ func NewBlockPuller(support consensus.ConsenterSupport,
...
 }
+
+// NewBlockFetcher creates a new block fetcher
+func NewBlockFetcher(support consensus.ConsenterSupport,
+	baseDialer *cluster.PredicateDialer,
+	clusterConfig localconfig.Cluster,
+	bccsp bccsp.BCCSP,
+) (BlockPuller, error) {
+	verifyBlockSequence := func(blocks []*common.Block, _ string) error {
+		return cluster.VerifyBlocks(blocks, support)
```
Comment (yacovm) [inline orderer/consensus/etcdraft/blockpuller.go:130]: we need to update the support as we pull blocks. We cannot re-use the support from the constructor.

File: orderer/consensus/etcdraft/blockpuller.go [hunk 5]
```diff
@@ -107,3 +119,108 @@ func NewBlockPuller(support consensus.ConsenterSupport,
...
+				Dialer:              stdDialer,
+				StopChannel:         make(chan struct{}),
+			}
+		},
+		Config:            fc,
+		Logger:            flogging.MustGetLogger("orderer.common.cluster.puller").With("channel", support.ChannelID()),
+		ShuffleTimeout:    shuffleTimeout,
+		LastShuffledAt:    time.Now(),
+		MaxByzantineNodes: maxByzantineNodes,
+		ConfirmByzantineBehavior: func(attestations []*orderer.BlockAttestation) bool {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/blockpuller.go:196]: This is more a question for the sake of discussion, but - where do we check the sequence of the attestation? What guarantees us that we receive the sequence we expect?

**Pull request merged**